pkg/credentialprovider/external.go

```diff
@@ -0,0 +1,228 @@
+	// first. For example, if for the given image "quay.io/coreos/etcd",
+	// credentials for "quay.io/coreos" should match before "quay.io".
+	sort.Sort(sort.Reverse(sort.StringSlice(keyring.index)))
+
+	// TODO remove
+	klog.V(2).Infof("Contents of keyring index %v", keyring.index)
+	klog.V(2).Infof("Contents of keyring providers %v", keyring.providers)
+
+	return keyring, nil
+}
+
+func (e *externalProviderKeyring) Lookup(image string) ([]AuthConfig, bool) {
+	var provider ExternalCredentialProvider
+	klog.V(2).Infof("Looking up %s with external provider.", image)
+
+	// range over the index as iterating over a map does not provide a predictable ordering
+	for _, matcher := range e.index {
+		// both k and image are schemeless URLs because even though schemes are allowed
+		// in the credential configurations, we remove them in Add.
+		if matched, _ := urlsMatchStr(matcher, image); matched {
+			provider = e.providers[matcher]
+		}
+	}
+
+	if provider == nil {
+		klog.Infof("No external provider available for the image: %s", image)
+		return []AuthConfig{}, false
+	}
+
+	cfg := provider.Provide(image)
+
+	providedConfigs := make([]AuthConfig, 0)
+	//for loc, ident := range cfg {
+	for _, ident := range cfg {
+		creds := AuthConfig{
+			Username: ident.Username,
+			Password: ident.Password,
+			Email:    ident.Email,
+		}
+		providedConfigs = append(providedConfigs, creds)
+	}
+	return providedConfigs, (len(providedConfigs) > 0)
+}
+
+type ExternalCredentialProvider interface {
+	Provide(image string) DockerConfig
+}
+
+type externalCredentialProvider struct {
+	command string
+	args    []string
+
+	// Each environment variable is in the form key=value
+	env []string
+
+	// Version of request/response structs for speaking to the plugin
+	apiVersion string
+}
+
+func (e *externalCredentialProvider) Provide(image string) DockerConfig {
+	cfg := make(DockerConfig)
+
+	response, err := e.request(image)
+	if err != nil {
+		klog.Errorf("Failed getting credential from external registry credential provider: %v", err)
+		return cfg
+	}
+
+	cfg[image] = DockerConfigEntry{
+		Username: *response.Username,
+		Password: *response.Password,
+	}
+
+	return cfg
+}
+
+func newRegistryCredentialYAMLEncoder(targetVersion schema.GroupVersion) (runtime.Encoder, error) {
+	mediaType := "application/yaml"
+	info, ok := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), mediaType)
+	if !ok {
+		return nil, fmt.Errorf("unsupported media type %q", mediaType)
+	}
+	return codecs.EncoderForVersion(info.Serializer, targetVersion), nil
+}
+
+func (e *externalCredentialProvider) request(image string) (*registrycredential.RegistryCredentialPluginResponse, error) {
+	var stdout, stderr, stdin bytes.Buffer
+	cmd := exec.Command(e.command, e.args...)
+	cmd.Stdout, cmd.Stderr, cmd.Stdin = &stdout, &stderr, &stdin
+
+	encoder, err := newRegistryCredentialYAMLEncoder(v1alpha1.SchemeGroupVersion)
```

Comment:
why YAML, wasn't it going to be JSON?

Reply:
Yeah apparently I can't remember the difference between the two, fixing.